File: src/lib/OPTIONS/hardware.cpp
#if !defined(UNIT_TEST)
#include "options.h"
#include "helpers.h"
#include "logging.h"
#include <LittleFS.h>
#include <ArduinoJson.h>

typedef enum {
    INT,
    BOOL,
    FLOAT,
    ARRAY,
    COUNT
} datatype_t;

// Every JSON key the hardware layout understands. The strings and the
// table itself live in flash (PROGMEM) because they are only walked at
// boot and when a layout is uploaded; on ESP8266 a plain const would
// otherwise cost ~3.3KB of DRAM.
#define HARDWARE_FIELDS(X) \
    X(HARDWARE_customised, "customised", BOOL) \
    X(HARDWARE_serial_rx, "serial_rx", INT) \
    X(HARDWARE_serial_tx, "serial_tx", INT) \
    X(HARDWARE_serial1_rx, "serial1_rx", INT) \
    X(HARDWARE_serial1_tx, "serial1_tx", INT) \
    X(HARDWARE_radio_busy, "radio_busy", INT) \
    X(HARDWARE_radio_busy_2, "radio_busy_2", INT) \
    X(HARDWARE_radio_dio0, "radio_dio0", INT) \
    X(HARDWARE_radio_dio0_2, "radio_dio0_2", INT) \
    X(HARDWARE_radio_dio1, "radio_dio1", INT) \
    X(HARDWARE_radio_dio1_2, "radio_dio1_2", INT) \
    X(HARDWARE_radio_miso, "radio_miso", INT) \
    X(HARDWARE_radio_mosi, "radio_mosi", INT) \
    X(HARDWARE_radio_nss, "radio_nss", INT) \
    X(HARDWARE_radio_nss_2, "radio_nss_2", INT) \
    X(HARDWARE_radio_rst, "radio_rst", INT) \
    X(HARDWARE_radio_rst_2, "radio_rst_2", INT) \
    X(HARDWARE_radio_sck, "radio_sck", INT) \
    X(HARDWARE_radio_dcdc, "radio_dcdc", BOOL) \
    X(HARDWARE_radio_rfo_hf, "radio_rfo_hf", BOOL) \
    X(HARDWARE_radio_rfsw_ctrl, "radio_rfsw_ctrl", ARRAY) \
    X(HARDWARE_radio_rfsw_ctrl_count, "radio_rfsw_ctrl", COUNT) \
    X(HARDWARE_radio_tcxo, "radio_tcxo", INT) \
    X(HARDWARE_radio_tcxo_delay, "radio_tcxo_delay", INT) \
    X(HARDWARE_ant_ctrl, "ant_ctrl", INT) \
    X(HARDWARE_ant_group, "ant_group", INT) \
    X(HARDWARE_power_enable, "power_enable", INT) \
    X(HARDWARE_power_apc2, "power_apc2", INT) \
    X(HARDWARE_power_rxen, "power_rxen", INT) \
    X(HARDWARE_power_txen, "power_txen", INT) \
    X(HARDWARE_power_rxen_2, "power_rxen_2", INT) \
    X(HARDWARE_power_txen_2, "power_txen_2", INT) \
    X(HARDWARE_power_lna_gain, "power_lna_gain", INT) \
    X(HARDWARE_power_min, "power_min", INT) \
    X(HARDWARE_power_max, "power_max", INT) \
    X(HARDWARE_power_default, "power_default", INT) \
    X(HARDWARE_power_pdet, "power_pdet", INT) \
    X(HARDWARE_power_pdet_intercept, "power_pdet_intercept", FLOAT) \
    X(HARDWARE_power_pdet_slope, "power_pdet_slope", FLOAT) \
    X(HARDWARE_power_control, "power_control", INT) \
    X(HARDWARE_power_values, "power_values", ARRAY) \
    X(HARDWARE_power_values_count, "power_values", COUNT) \
    X(HARDWARE_power_values2, "power_values2", ARRAY) \
    X(HARDWARE_power_values_dual, "power_values_dual", ARRAY) \
    X(HARDWARE_power_values_dual_count, "power_values_dual", COUNT) \
    X(HARDWARE_joystick, "joystick", INT) \
    X(HARDWARE_joystick_values, "joystick_values", ARRAY) \
    X(HARDWARE_five_way1, "five_way1", INT) \
    X(HARDWARE_five_way2, "five_way2", INT) \
    X(HARDWARE_five_way3, "five_way3", INT) \
    X(HARDWARE_button, "button", INT) \
    X(HARDWARE_button_led_index, "button_led_index", INT) \
    X(HARDWARE_button2, "button2", INT) \
    X(HARDWARE_button2_led_index, "button2_led_index", INT) \
    X(HARDWARE_led, "led", INT) \
    X(HARDWARE_led_blue, "led_blue", INT) \
    X(HARDWARE_led_blue_invert, "led_blue_invert", BOOL) \
    X(HARDWARE_led_green, "led_green", INT) \
    X(HARDWARE_led_green_invert, "led_green_invert", BOOL) \
    X(HARDWARE_led_green_red, "led_green_red", INT) \
    X(HARDWARE_led_red, "led_red", INT) \
    X(HARDWARE_led_red_invert, "led_red_invert", BOOL) \
    X(HARDWARE_led_red_green, "led_red_green", INT) \
    X(HARDWARE_led_rgb, "led_rgb", INT) \
    X(HARDWARE_led_rgb_isgrb, "led_rgb_isgrb", BOOL) \
    X(HARDWARE_ledidx_rgb_status, "ledidx_rgb_status", ARRAY) \
    X(HARDWARE_ledidx_rgb_status_count, "ledidx_rgb_status", COUNT) \
    X(HARDWARE_ledidx_rgb_vtx, "ledidx_rgb_vtx", ARRAY) \
    X(HARDWARE_ledidx_rgb_vtx_count, "ledidx_rgb_vtx", COUNT) \
    X(HARDWARE_ledidx_rgb_boot, "ledidx_rgb_boot", ARRAY) \
    X(HARDWARE_ledidx_rgb_boot_count, "ledidx_rgb_boot", COUNT) \
    X(HARDWARE_screen_cs, "screen_cs", INT) \
    X(HARDWARE_screen_dc, "screen_dc", INT) \
    X(HARDWARE_screen_mosi, "screen_mosi", INT) \
    X(HARDWARE_screen_rst, "screen_rst", INT) \
    X(HARDWARE_screen_sck, "screen_sck", INT) \
    X(HARDWARE_screen_sda, "screen_sda", INT) \
    X(HARDWARE_screen_type, "screen_type", INT) \
    X(HARDWARE_screen_reversed, "screen_reversed", BOOL) \
    X(HARDWARE_screen_mirror, "screen_mirror", BOOL) \
    X(HARDWARE_screen_bl, "screen_bl", INT) \
    X(HARDWARE_use_backpack, "use_backpack", BOOL) \
    X(HARDWARE_debug_backpack_baud, "debug_backpack_baud", INT) \
    X(HARDWARE_debug_backpack_rx, "debug_backpack_rx", INT) \
    X(HARDWARE_debug_backpack_tx, "debug_backpack_tx", INT) \
    X(HARDWARE_backpack_boot, "backpack_boot", INT) \
    X(HARDWARE_backpack_en, "backpack_en", INT) \
    X(HARDWARE_passthrough_baud, "passthrough_baud", INT) \
    X(HARDWARE_i2c_scl, "i2c_scl", INT) \
    X(HARDWARE_i2c_sda, "i2c_sda", INT) \
    X(HARDWARE_misc_gsensor_int, "misc_gsensor_int", INT) \
    X(HARDWARE_misc_buzzer, "misc_buzzer", INT) \
    X(HARDWARE_misc_fan_en, "misc_fan_en", INT) \
    X(HARDWARE_misc_fan_pwm, "misc_fan_pwm", INT) \
    X(HARDWARE_misc_fan_tacho, "misc_fan_tacho", INT) \
    X(HARDWARE_misc_fan_speeds, "misc_fan_speeds", ARRAY) \
    X(HARDWARE_misc_fan_speeds_count, "misc_fan_speeds", COUNT) \
    X(HARDWARE_gsensor_stk8xxx, "gsensor_stk8xxx", BOOL) \
    X(HARDWARE_thermal_lm75a, "thermal_lm75a", BOOL) \
    X(HARDWARE_pwm_outputs, "pwm_outputs", ARRAY) \
    X(HARDWARE_pwm_outputs_count, "pwm_outputs", COUNT) \
    X(HARDWARE_pwm_out_only, "pwm_out_only", BOOL) \
    X(HARDWARE_vbat, "vbat", INT) \
    X(HARDWARE_vbat_offset, "vbat_offset", INT) \
    X(HARDWARE_vbat_scale, "vbat_scale", INT) \
    X(HARDWARE_vbat_atten, "vbat_atten", INT) \
    X(HARDWARE_vbat_noreading, "vbat_noreading", INT) \
    X(HARDWARE_vbat_cal_min, "vbat_cal_min", INT) \
    X(HARDWARE_vbat_cal_max, "vbat_cal_max", INT) \
    X(HARDWARE_vsrc1, "vsrc1", INT) \
    X(HARDWARE_vsrc1_offset, "vsrc1_offset", INT) \
    X(HARDWARE_vsrc1_scale, "vsrc1_scale", INT) \
    X(HARDWARE_vsrc1_atten, "vsrc1_atten", INT) \
    X(HARDWARE_vsrc1_noreading, "vsrc1_noreading", INT) \
    X(HARDWARE_vsrc1_cal_min, "vsrc1_cal_min", INT) \
    X(HARDWARE_vsrc1_cal_max, "vsrc1_cal_max", INT) \
    X(HARDWARE_vsrc2, "vsrc2", INT) \
    X(HARDWARE_vsrc2_offset, "vsrc2_offset", INT) \
    X(HARDWARE_vsrc2_scale, "vsrc2_scale", INT) \
    X(HARDWARE_vsrc2_atten, "vsrc2_atten", INT) \
    X(HARDWARE_vsrc2_noreading, "vsrc2_noreading", INT) \
    X(HARDWARE_vsrc2_cal_min, "vsrc2_cal_min", INT) \
    X(HARDWARE_vsrc2_cal_max, "vsrc2_cal_max", INT) \
    X(HARDWARE_vsrc3, "vsrc3", INT) \
    X(HARDWARE_vsrc3_offset, "vsrc3_offset", INT) \
    X(HARDWARE_vsrc3_scale, "vsrc3_scale", INT) \
    X(HARDWARE_vsrc3_atten, "vsrc3_atten", INT) \
    X(HARDWARE_vsrc3_noreading, "vsrc3_noreading", INT) \
    X(HARDWARE_vsrc3_cal_min, "vsrc3_cal_min", INT) \
    X(HARDWARE_vsrc3_cal_max, "vsrc3_cal_max", INT) \
    X(HARDWARE_vtx_amp_pwm, "vtx_amp_pwm", INT) \
    X(HARDWARE_vtx_amp_vpd, "vtx_amp_vpd", INT) \
    X(HARDWARE_vtx_amp_vref, "vtx_amp_vref", INT) \
    X(HARDWARE_vtx_nss, "vtx_nss", INT) \
    X(HARDWARE_vtx_miso, "vtx_miso", INT) \
    X(HARDWARE_vtx_mosi, "vtx_mosi", INT) \
    X(HARDWARE_vtx_sck, "vtx_sck", INT) \
    X(HARDWARE_vtx_amp_vpd_25mW, "vtx_amp_vpd_25mW", ARRAY) \
    X(HARDWARE_vtx_amp_vpd_100mW, "vtx_amp_vpd_100mW", ARRAY) \
    X(HARDWARE_vtx_amp_pwm_25mW, "vtx_amp_pwm_25mW", ARRAY) \
    X(HARDWARE_vtx_amp_pwm_100mW, "vtx_amp_pwm_100mW", ARRAY)

#define X(pos, str, typ) static const char name_##pos[] PROGMEM = str;
HARDWARE_FIELDS(X)
#undef X

typedef struct {
    nameType position;
    const char *name;
    datatype_t type;
} field_t;

static const field_t fields[] PROGMEM = {
#define X(pos, str, typ) {pos, name_##pos, typ},
HARDWARE_FIELDS(X)
#undef X
};

typedef union {
    int int_value;
    bool bool_value;
    float float_value;
    int16_t *array_value;
} data_holder_t;

static data_holder_t hardware[HARDWARE_LAST];
static String builtinHardwareConfig;

String& getHardware()
{
    File file = LittleFS.open("/hardware.json", "r");
    if (!file || file.isDirectory())
    {
        if (file)
        {
            file.close();
        }
        // Try JSON at the end of the firmware
        return builtinHardwareConfig;
    }
    builtinHardwareConfig = file.readString();
    return builtinHardwareConfig;
}

static void hardware_ClearAllFields()
{
    for (size_t i = 0; i < ARRAY_SIZE(fields); i++) {
        field_t field;
        memcpy_P(&field, &fields[i], sizeof(field));
        switch (field.type) {
            case INT:
                hardware[field.position].int_value = -1;
                break;
            case BOOL:
                hardware[field.position].bool_value = false;
                break;
            case FLOAT:
                hardware[field.position].float_value = 0.0;
                break;
            case ARRAY:
                hardware[field.position].array_value = nullptr;
                break;
            case COUNT:
                hardware[field.position].int_value = 0;
                break;
        }
    }
}

static void hardware_LoadFieldsFromDoc(JsonDocument &doc)
{
    for (size_t i = 0; i < ARRAY_SIZE(fields); i++) {
        field_t field;
        memcpy_P(&field, &fields[i], sizeof(field));
        const __FlashStringHelper *name = FPSTR(field.name);
        if (doc[name].is<JsonVariant>()) {
            switch (field.type) {
                case INT:
                    hardware[field.position].int_value = doc[name];
                    break;
                case BOOL:
                    hardware[field.position].bool_value = doc[name];
                    break;
                case FLOAT:
                    hardware[field.position].float_value = doc[name];
                    break;
                case ARRAY:
                    {
                        JsonArray array = doc[name].as<JsonArray>();
                        hardware[field.position].array_value = new int16_t[array.size()];
                        copyArray(array, hardware[field.position].array_value, array.size());
                    }
                    break;
                case COUNT:
                    {
                        JsonArray array = doc[name].as<JsonArray>();
                        hardware[field.position].int_value = (int)array.size();
                    }
                    break;
            }
        }
    }
}

bool hardware_init(EspFlashStream &strmFlash)
{
    hardware_ClearAllFields();
    builtinHardwareConfig.clear();

    Stream *strmSrc;
    JsonDocument doc;
    File file = LittleFS.open("/hardware.json", "r");
    if (!file || file.isDirectory()) {
        constexpr size_t hardwareConfigOffset = ELRSOPTS_PRODUCTNAME_SIZE + ELRSOPTS_DEVICENAME_SIZE + ELRSOPTS_OPTIONS_SIZE;
        strmFlash.setPosition(hardwareConfigOffset);
        if (!options_HasStringInFlash(strmFlash))
        {
            return false;
        }

        strmSrc = &strmFlash;
    }
    else
    {
        strmSrc = &file;
    }

    DeserializationError error = deserializeJson(doc, *strmSrc);
    if (error)
    {
        return false;
    }
    serializeJson(doc, builtinHardwareConfig);

    hardware_LoadFieldsFromDoc(doc);

    return true;
}

int hardware_pin(nameType name)
{
    return hardware[name].int_value;
}

bool hardware_flag(nameType name)
{
    return hardware[name].bool_value;
}

int hardware_int(nameType name)
{
    return hardware[name].int_value;
}

float hardware_float(nameType name)
{
    return hardware[name].float_value;
}

const int16_t* hardware_i16_array(nameType name)
{
    return hardware[name].array_value;
}

const uint16_t* hardware_u16_array(nameType name)
{
    return (uint16_t *)hardware[name].array_value;
}
#endif

File: src/hardware/.github/hardware.py
from enum import Enum
from pin_functions import get_pin_function


class FieldType(Enum):
    INT = 1
    BOOL = 2
    FLOAT = 3
    ARRAY = 4
    PIN = 100   # marker
    INPUT = 101
    OUTPUT = 102
    BIDIR = 103
    ADC = 104


hardware_fields = {
    "serial_rx": FieldType.INPUT,
    "serial_tx": FieldType.OUTPUT,
    "serial1_rx": FieldType.INPUT,
    "serial1_tx": FieldType.OUTPUT,
    "radio_busy": FieldType.INPUT,
    "radio_busy_2": FieldType.INPUT,
    "radio_dio0": FieldType.INPUT,
    "radio_dio0_2": FieldType.INPUT,
    "radio_dio1": FieldType.INPUT,
    "radio_dio1_2": FieldType.INPUT,
    "radio_dio2": FieldType.INPUT,
    "radio_miso": FieldType.INPUT,
    "radio_mosi": FieldType.OUTPUT,
    "radio_nss": FieldType.OUTPUT,
    "radio_nss_2": FieldType.OUTPUT,
    "radio_rst": FieldType.OUTPUT,
    "radio_rst_2": FieldType.OUTPUT,
    "radio_sck": FieldType.OUTPUT,
    "radio_dcdc": FieldType.BOOL,
    "radio_rfo_hf": FieldType.BOOL,
    "ant_ctrl": FieldType.OUTPUT,
    "ant_ctrl_compl": FieldType.OUTPUT,
    "power_enable": FieldType.OUTPUT,
    "power_apc1": FieldType.OUTPUT,
    "power_apc2": FieldType.OUTPUT,
    "power_rxen": FieldType.OUTPUT,
    "power_txen": FieldType.OUTPUT,
    "power_rxen_2": FieldType.OUTPUT,
    "power_txen_2": FieldType.OUTPUT,
    "power_lna_gain": FieldType.INT,
    "power_min": FieldType.INT,
    "power_high": FieldType.INT,
    "power_max": FieldType.INT,
    "power_default": FieldType.INT,
    "power_pdet": FieldType.ADC,
    "power_pdet_intercept": FieldType.FLOAT,
    "power_pdet_slope": FieldType.FLOAT,
    "power_control": FieldType.INT,
    "power_values": FieldType.ARRAY,
    "power_values2": FieldType.ARRAY,
    "power_values_dual": FieldType.ARRAY,
    "joystick": FieldType.ADC,
    "joystick_values": FieldType.ARRAY,
    "five_way1": FieldType.INPUT,
    "five_way2": FieldType.INPUT,
    "five_way3": FieldType.INPUT,
    "button": FieldType.INPUT,
    "button_led_index": FieldType.INT,
    "button2": FieldType.INPUT,
    "button2_led_index": FieldType.INT,
    "led": FieldType.OUTPUT,
    "led_blue": FieldType.OUTPUT,
    "led_blue_invert": FieldType.BOOL,
    "led_green": FieldType.OUTPUT,
    "led_green_invert": FieldType.BOOL,
    "led_green_red": FieldType.OUTPUT,
    "led_red": FieldType.OUTPUT,
    "led_red_invert": FieldType.BOOL,
    "led_red_green": FieldType.OUTPUT,
    "led_rgb": FieldType.OUTPUT,
    "led_rgb_isgrb": FieldType.BOOL,
    "ledidx_rgb_status": FieldType.ARRAY,
    "ledidx_rgb_vtx": FieldType.ARRAY,
    "ledidx_rgb_boot": FieldType.ARRAY,
    "screen_cs": FieldType.OUTPUT,
    "screen_dc": FieldType.OUTPUT,
    "screen_mosi": FieldType.OUTPUT,
    "screen_rst": FieldType.OUTPUT,
    "screen_sck": FieldType.OUTPUT,
    "screen_sda": FieldType.OUTPUT,
    "screen_type": FieldType.INT,
    "screen_reversed": FieldType.BOOL,
    "screen_bl": FieldType.OUTPUT,
    "use_backpack": FieldType.BOOL,
    "debug_backpack_baud": FieldType.INT,
    "debug_backpack_rx": FieldType.INPUT,
    "debug_backpack_tx": FieldType.OUTPUT,
    "backpack_boot": FieldType.OUTPUT,
    "backpack_en": FieldType.OUTPUT,
    "passthrough_baud": FieldType.INT,
    "i2c_scl": FieldType.OUTPUT,
    "i2c_sda": FieldType.BIDIR,
    "misc_gsensor_int": FieldType.INPUT,
    "misc_buzzer": FieldType.OUTPUT,
    "misc_fan_en": FieldType.OUTPUT,
    "misc_fan_pwm": FieldType.OUTPUT,
    "misc_fan_tacho": FieldType.INPUT,
    "misc_fan_speeds": FieldType.ARRAY,
    "gsensor_stk8xxx": FieldType.BOOL,
    "thermal_lm75a": FieldType.BOOL,
    "pwm_outputs": FieldType.ARRAY,
    "vbat": FieldType.ADC,
    "vbat_offset": FieldType.INT,
    "vbat_scale": FieldType.INT,
    "vbat_atten": FieldType.INT,
    "vtx_amp_pwm": FieldType.OUTPUT,
    "vtx_amp_vpd": FieldType.ADC,
    "vtx_amp_vref": FieldType.OUTPUT,
    "vtx_nss": FieldType.OUTPUT,
    "vtx_miso": FieldType.INPUT,
    "vtx_mosi": FieldType.OUTPUT,
    "vtx_sck": FieldType.OUTPUT,
    "vtx_amp_vpd_25mW": FieldType.ARRAY,
    "vtx_amp_vpd_100mW": FieldType.ARRAY,
    "vtx_amp_pwm_25mW": FieldType.ARRAY,
    "vtx_amp_pwm_100mW": FieldType.ARRAY,
    "ir_transponder": FieldType.OUTPUT,
    "gyro_nss": FieldType.OUTPUT,
    "gyro_miso": FieldType.INPUT,
    "gyro_mosi": FieldType.OUTPUT,
    "gyro_sck": FieldType.OUTPUT,
    "gyro_int": FieldType.INPUT,
    "adc_a1": FieldType.ADC,
    "adc_a2": FieldType.ADC
}

field_groups = {
    "all": [
        # if one of the first group then all the first and second groups and
        # at-least one of the third group must also be defined
        [["serial_rx", "serial_tx"], [], []],
        [["serial1_rx", "serial1_tx"], [], []],
        [["power_min", "power_high", "power_max", "power_default", "power_control", "power_values"], [], []],
        [["debug_backpack_baud", "debug_backpack_rx", "debug_backpack_tx"], [], []],
        [["use_backpack"], ["debug_backpack_baud", "debug_backpack_rx", "debug_backpack_tx"], []],
        [["backpack_boot", "backpack_en"], ["use_backpack", "debug_backpack_baud", "debug_backpack_rx", "debug_backpack_tx"], []],
        [["i2c_scl", "i2c_sda"], [], []],
        [["joystick", "joystick_values"], [], []],
        [["five_way1", "five_way2", "five_way3"], [], []],
        [["misc_fan_pwm", "misc_fan_speeds"], [], []],
        [["vbat", "vbat_offset", "vbat_scale"], [], []],
        [["power_pdet", "power_pdet_intercept", "power_pdet_slope"], [], []],
        [["screen_sda"], ["screen_sck", "screen_type"], []],
        [["screen_cs", "screen_dc", "screen_mosi"], ["screen_type", "screen_sck", "screen_rst"], []],
        [["vtx_amp_pwm", "vtx_amp_vpd", "vtx_amp_vref", "vtx_nss", "vtx_miso", "vtx_mosi", "vtx_sck", "vtx_amp_vpd_25mW", "vtx_amp_vpd_100mW"], [], []],
        [["gyro_nss", "gyro_miso", "gyro_mosi", "gyro_sck", "gyro_int"], [], []]
    ],
    "2400": [
        [["radio_dio1", "radio_miso", "radio_mosi", "radio_sck", "radio_nss"], [], ["radio_rst", "pwm_outputs"]],
        [["radio_busy"], ["radio_dio1", "radio_miso", "radio_mosi", "radio_sck", "radio_nss"], ["radio_rst", "pwm_outputs"]],
        [["radio_rst"], ["radio_dio1", "radio_miso", "radio_mosi", "radio_sck", "radio_nss"], []],
        [["radio_dio1_2", "radio_nss_2"], ["radio_miso", "radio_mosi", "radio_sck"], ["radio_rst_2", "radio_rst", "pwm_outputs"]],
        [["radio_busy_2"], ["radio_dio1_2", "radio_miso", "radio_mosi", "radio_sck", "radio_nss_2"], ["radio_rst_2", "radio_rst", "pwm_outputs"]],
        [["radio_rst_2"], ["radio_dio1_2", "radio_miso", "radio_mosi", "radio_sck", "radio_nss_2"], []]
    ],
    "900": [
        [["radio_dio0", "radio_miso", "radio_mosi", "radio_sck", "radio_nss"], [], ["radio_rst", "pwm_outputs"]],
        [["radio_rst"], ["radio_dio0", "radio_miso", "radio_mosi", "radio_sck", "radio_nss"], []],
        [["radio_dio0_2", "radio_nss_2"], ["radio_miso", "radio_mosi", "radio_sck"], ["radio_rst_2", "radio_rst", "pwm_outputs"]],
        [["radio_rst_2"], ["radio_dio0_2", "radio_miso", "radio_mosi", "radio_sck", "radio_nss_2"], []]
    ],
    "dual": [
        [["radio_dio1", "radio_miso", "radio_mosi", "radio_sck", "radio_nss"], [], ["radio_rst", "pwm_outputs"]],
        [["radio_busy"], ["radio_dio1", "radio_miso", "radio_mosi", "radio_sck", "radio_nss"], ["radio_rst", "pwm_outputs"]],
        [["radio_rst"], ["radio_dio1", "radio_miso", "radio_mosi", "radio_sck", "radio_nss"], []],
        [["radio_dio1_2", "radio_nss_2"], ["radio_miso", "radio_mosi", "radio_sck"], ["radio_rst_2", "radio_rst", "pwm_outputs"]],
        [["radio_busy_2"], ["radio_dio1_2", "radio_miso", "radio_mosi", "radio_sck", "radio_nss_2"], ["radio_rst_2", "radio_rst", "pwm_outputs"]],
        [["radio_rst_2"], ["radio_dio1_2", "radio_miso", "radio_mosi", "radio_sck", "radio_nss_2"], []]
    ]
}

mutually_exclusive_groups = [
    ["ant_ctrl", "ant_ctrl_compl"]
]

allowable_duplicates = [
    ['serial_rx', 'serial_tx'],
    ['screen_sck', 'i2c_scl'],
    ['screen_sda', 'screen_mosi', 'i2c_sda']
]
allowable_pwm_shared = [
    'serial_rx', 'serial_tx', 'serial1_rx', 'serial1_tx', 'i2c_scl', 'i2c_sda'
]

used_pins = {}


def ignore_undef_pins(pair):
    key, value = pair
    # ignore keys that start with a comment character
    if key[0] in '/#;':
        return False

    # FIXME, <0 should really be -1, but I used -pin number to mean inverted for the backlight
    if key in hardware_fields and hardware_fields[key].value > FieldType.PIN.value and value < 0:
        return False
    else:
        return True


def validate(target, layout, device, type):
    global used_pins
    had_error = False
    used_pins = {}
    for field in dict(filter(ignore_undef_pins, layout.items())):
        # Ensure that the layout field is a valid field from the hardware list
        if field not in hardware_fields.keys():
            print(f'device "{target}" has an unknown field name {field}')
            had_error = True
        else:
            had_error |= validate_pin_uniqueness(target, layout, field)
            had_error |= validate_grouping(target, layout, field, type)
            had_error |= validate_pin_function(target, layout, field, device['platform'])
    had_error |= validate_power_config(target, layout)
    had_error |= validate_backpack(target, layout)
    had_error |= validate_joystick(target, layout)
    had_error |= validate_pwm_outputs(target, layout)
    had_error |= validate_vtx_amp_pwm(target, layout, device['platform'])
    return had_error


def validate_grouping(target, layout, field, radio):
    had_error = validate_field_grouping(target, layout, field, 'all')
    if radio.endswith('_2400'):
        had_error |= validate_field_grouping(target, layout, field, '2400')
    elif radio.endswith('_900'):
        had_error |= validate_field_grouping(target, layout, field, '900')
    elif radio.endswith('_dual'):
        had_error |= validate_field_grouping(target, layout, field, 'dual')
    return had_error


def validate_field_grouping(target, layout, field, field_group):
    had_error = False
    for group in field_groups[field_group]:
        if field in group[0]:
            for must in group[0] + group[1]:
                if must not in layout:
                    print(f'device "{target}" because "{field}" is defined all other related fields must also be defined {must}')
                    print(f'\t{group[0] + group[1]}')
                    had_error = True
            found = True if group[2] == [] else False
            for one in group[2]:
                if one in layout:
                    found = True
            if not found:
                print(f'device "{target}" because "{field}" is defined at least one of the following fields must also be {group[2]}')
                had_error = True
    return had_error


def validate_pin_uniqueness(target, layout, field):
    had_error = False
    if hardware_fields[field].value > FieldType.PIN.value:
        pin = layout[field]
        if pin in used_pins.keys():
            allowed = False
            for duplicate in allowable_duplicates:
                if field in duplicate and used_pins[pin] in duplicate:
                    allowed = True
            if not allowed:
                print(f'device "{target}" PIN {pin} "{field}" is already assigned to "{used_pins[pin]}"')
                had_error = True
        else:
            used_pins[pin] = field
    return had_error


def validate_power_config(target, layout):
    had_error = False
    if 'power_values' in layout:
        power_values = layout['power_values']
        power_max = layout['power_max']
        power_min = layout['power_min']
        power_default = layout['power_default']
        power_high = layout['power_high']
        if power_min > power_max:
            print(f'device "{target}" power_min must be less than or equal to power_max')
            had_error = True
        if power_default < power_min or power_default > power_max:
            print(f'device "{target}" power_default must lie between power_min and power_max')
            had_error = True
        if power_high < power_min or power_high > power_max:
            print(f'device "{target}" power_high must lie between power_min and power_max')
            had_error = True
        if power_max - power_min + 1 != len(power_values):
            print(f'device "{target}" power_values must have the correct number of entries to match all values from power_min to power_max')
            had_error = power_max - power_min + 1 > len(power_values)
        if layout['power_control'] == 3 and 'power_apc2' not in layout:
            print(f'device "{target}" defines power_control as DACWRITE and power_apc2 is undefined')
            had_error = True
        if 'power_values2' in layout:
            if len(layout['power_values2']) != len(power_values):
                print(f'device "{target}" power_values2 must have the same number of entries as power_values')
                had_error = True
            if layout['power_control'] != 3:
                print(f'device "{target}" power_values2 is defined so power_control must be set to 3 (DACWRITE)')
                had_error = True
            if 'power_apc2' not in layout:
                print(f'device "{target}" power_values2 is defined so the power_apc2 pin must also be defined')
                had_error = True
    return had_error


def validate_backpack(target, layout):
    had_error = False
    if 'passthrough_baud' in layout:
        if layout['serial_rx'] == layout['serial_tx'] and layout['passthrough_baud'] != 230400:
            print(f'device "{target}" an external module with a backpack should set the baud rate to 230400')
            had_error = True
        if layout['serial_rx'] != layout['serial_tx'] and layout['passthrough_baud'] != 460800:
            print(f'device "{target}" an internal module with a backpack should set the baud rate to 460800')
            had_error = True
    return had_error


def validate_joystick(target, layout):
    had_error = False
    if 'joystick' in layout or 'joystick_values' in layout:
        if 'joystick' not in layout:
            print(f'device "{target}" joystick_values is defined so the joystick pin must also be defined')
            had_error = True
        elif 'joystick_values' not in layout:
            print(f'device "{target}" joystick is defined so the joystick_values must also be defined')
            had_error = True
        elif len(layout['joystick_values']) != 6:
            print(f'device "{target}" joystick_values must have 6 values defined')
            had_error = True
        else:
            for value in layout['joystick_values']:
                if value < 0 or value > 4095:
                    print(f'device "{target}" joystick_values must be between 0 and 4095 inclusive')
                    had_error = True
    return had_error


def validate_pwm_outputs(target, layout):
    had_error = False
    if 'pwm_outputs' in layout:
        for field in hardware_fields:
            if hardware_fields[field].value > FieldType.PIN.value and \
                    field in layout and \
                    layout[field] in layout['pwm_outputs'] and \
                    field not in allowable_pwm_shared:
                print(f'device "{target}" pwm_output pin {layout[field]} is not allowed to be shared with {field}')
                had_error = True
    return had_error


def validate_pin_function(target, layout, field, platform):
    if hardware_fields[field].value > FieldType.PIN.value:
        function = get_pin_function(platform, layout[field])
        if function is None:
            print(f'device "{target}" has an invalid pin number for {field}, {layout[field]}')
            return True
        if hardware_fields[field] == FieldType.INPUT and not (function & 1):
            print(f'device "{target}" pin for {field} must be assigned to a pin that supports INPUT')
            return True
    return False


def validate_vtx_amp_pwm(target, layout, platform):
    if platform == 'esp32' and 'vtx_amp_pwm' in layout:
        function = get_pin_function(platform, layout['vtx_amp_pwm'])
        if function is not None and function & 8 != 8:
            print(f'device "{target}" "vtx_amp_pwm" is preferred to be a DAC pin if possible, but PWM output is supported')
    return False
